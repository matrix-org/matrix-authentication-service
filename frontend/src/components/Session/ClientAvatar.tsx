// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { CSSProperties } from "react";

import styles from "./ClientAvatar.module.css";

/**
 * Render a client logo avatar when logoUri is truthy
 * Otherwise return null
 */
const ClientAvatar: React.FC<{
  name: string;
  logoUri?: string;
  size: string;
}> = ({ name, logoUri, size }) => {
  // compound's lazy loading for avatars does not allow CORS requests
  // so use our own avatar styled img
  if (logoUri) {
    return (
      <img
        className={styles.avatar}
        src={logoUri}
        alt={name}
        style={
          {
            "--mas-avatar-size": size,
          } as CSSProperties
        }
      />
    );
  }
  return null;
};

export default ClientAvatar;
