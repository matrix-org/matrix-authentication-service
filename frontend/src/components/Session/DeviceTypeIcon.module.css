/* Copyright 2023 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.icon {
    color: var(--cpd-color-icon-primary);
    background-color: var(--cpd-color-bg-subtle-primary);
    flex: 0 0 var(--cpd-space-4x);
    box-sizing: content-box;
    height: var(--cpd-space-4x);
    width: var(--cpd-space-4x);
    padding: var(--cpd-space-2x);
    border-radius: var(--cpd-space-2x);
}