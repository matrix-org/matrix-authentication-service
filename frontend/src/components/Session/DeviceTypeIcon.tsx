// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import IconComputer from "@vector-im/compound-design-tokens/icons/computer.svg?react";
import IconMobile from "@vector-im/compound-design-tokens/icons/mobile.svg?react";
import IconUnknown from "@vector-im/compound-design-tokens/icons/unknown.svg?react";
import IconBrowser from "@vector-im/compound-design-tokens/icons/web-browser.svg?react";
import { FunctionComponent, SVGProps } from "react";

import { DeviceType } from "../../utils/parseUserAgent";

import styles from "./DeviceTypeIcon.module.css";

const deviceTypeToIcon: Record<
  DeviceType,
  FunctionComponent<SVGProps<SVGSVGElement> & { title?: string | undefined }>
> = {
  [DeviceType.Unknown]: IconUnknown,
  [DeviceType.Desktop]: IconComputer,
  [DeviceType.Mobile]: IconMobile,
  [DeviceType.Web]: IconBrowser,
};

const deviceTypeToLabel: Record<DeviceType, string> = {
  [DeviceType.Unknown]: "Unknown device type",
  [DeviceType.Desktop]: "Desktop",
  [DeviceType.Mobile]: "Mobile",
  [DeviceType.Web]: "Web",
};

const DeviceTypeIcon: React.FC<{ deviceType: DeviceType }> = ({
  deviceType,
}) => {
  const Icon = deviceTypeToIcon[deviceType];
  const label = deviceTypeToLabel[deviceType];

  return <Icon className={styles.icon} aria-label={label} />;
};

export default DeviceTypeIcon;
