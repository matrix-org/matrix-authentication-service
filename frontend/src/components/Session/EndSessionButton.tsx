// Copyright 2023, 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import IconSignOut from "@vector-im/compound-design-tokens/assets/web/icons/sign-out";
import { Button } from "@vector-im/compound-web";
import { useState } from "react";
import { useTranslation } from "react-i18next";

import * as Dialog from "../Dialog";
import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";

/**
 * Generic end session button
 * Handles loading state while endSession is in progress
 */
const EndSessionButton: React.FC<
  React.PropsWithChildren<{ endSession: () => Promise<void> }>
> = ({ children, endSession }) => {
  const [inProgress, setInProgress] = useState(false);
  const [open, setOpen] = useState(false);
  const { t } = useTranslation();

  const onConfirm = async (
    e: React.MouseEvent<HTMLButtonElement>,
  ): Promise<void> => {
    e.preventDefault();

    setInProgress(true);
    try {
      await endSession();
      setOpen(false);
    } catch (error) {
      console.error("Failed to end session", error);
    }
    setInProgress(false);
  };

  return (
    <Dialog.Dialog
      open={open}
      onOpenChange={setOpen}
      trigger={
        <Button kind="secondary" destructive size="sm" Icon={IconSignOut}>
          {t("frontend.end_session_button.text")}
        </Button>
      }
    >
      <Dialog.Title>
        {t("frontend.end_session_button.confirmation_modal_title")}
      </Dialog.Title>

      {children && <Dialog.Description>{children}</Dialog.Description>}

      <Button
        type="button"
        kind="primary"
        destructive
        onClick={onConfirm}
        disabled={inProgress}
        Icon={inProgress ? undefined : IconSignOut}
      >
        {inProgress && <LoadingSpinner inline />}
        {t("frontend.end_session_button.text")}
      </Button>

      <Dialog.Close asChild>
        <Button kind="tertiary">{t("action.cancel")}</Button>
      </Dialog.Close>
    </Dialog.Dialog>
  );
};

export default EndSessionButton;
