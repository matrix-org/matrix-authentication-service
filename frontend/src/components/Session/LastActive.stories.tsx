// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { Meta, StoryObj } from "@storybook/react";
import { parseISO, subDays, subHours } from "date-fns";

import LastActive from "./LastActive";

const meta = {
  title: "UI/Session/Last active time",
  component: LastActive,
  argTypes: {
    lastActive: { control: { type: "date" } },
    now: { control: { type: "date" } },
  },
  tags: ["autodocs"],
} satisfies Meta<typeof LastActive>;

export default meta;
type Story = StoryObj<typeof LastActive>;

const now = parseISO("2023-09-18T01:12:00.000Z");

export const Basic: Story = {
  args: {
    // An hour ago
    lastActive: subHours(now, 1),
    now,
  },
};

export const ActiveThreeDaysAgo: Story = {
  args: {
    // Three days ago
    lastActive: subDays(now, 3),
    now,
  },
};

export const ActiveNow: Story = {
  args: {
    lastActive: now,
    now,
  },
};

export const Inactive: Story = {
  args: {
    // 91 days ago
    lastActive: subDays(now, 91),
    now,
  },
};
