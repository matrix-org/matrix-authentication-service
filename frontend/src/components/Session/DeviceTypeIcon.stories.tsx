// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { Meta, StoryObj } from "@storybook/react";

import { DeviceType } from "../../gql/graphql";

import DeviceTypeIcon from "./DeviceTypeIcon";

const meta = {
  title: "UI/Session/Device Type Icon",
  component: DeviceTypeIcon,
  tags: ["autodocs"],
  args: {
    deviceType: DeviceType.Unknown,
  },
  argTypes: {
    deviceType: {
      control: "select",
      options: [
        DeviceType.Unknown,
        DeviceType.Pc,
        DeviceType.Mobile,
        DeviceType.Tablet,
      ],
    },
  },
} satisfies Meta<typeof DeviceTypeIcon>;

export default meta;
type Story = StoryObj<typeof DeviceTypeIcon>;

export const Unknown: Story = {};

export const Pc: Story = {
  args: {
    deviceType: DeviceType.Pc,
  },
};
export const Mobile: Story = {
  args: {
    deviceType: DeviceType.Mobile,
  },
};
export const Tablet: Story = {
  args: {
    deviceType: DeviceType.Tablet,
  },
};
