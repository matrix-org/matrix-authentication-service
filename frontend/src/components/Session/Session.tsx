// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { H6, Body, Badge } from "@vector-im/compound-web";
import { ReactNode } from "react";
import { Trans, useTranslation } from "react-i18next";

import { DeviceType } from "../../utils/parseUserAgent";
import Block from "../Block";
import DateTime from "../DateTime";

import ClientAvatar from "./ClientAvatar";
import DeviceTypeIcon from "./DeviceTypeIcon";
import LastActive from "./LastActive";
import styles from "./Session.module.css";

const SessionMetadata: React.FC<React.ComponentProps<typeof Body>> = (
  props,
) => <Body {...props} size="sm" className={styles.sessionMetadata} />;

type SessionProps = {
  id: string;
  name?: string | ReactNode;
  createdAt: Date;
  finishedAt?: Date;
  clientName?: string;
  clientLogoUri?: string;
  isCurrent?: boolean;
  deviceType?: DeviceType;
  lastActiveIp?: string;
  lastActiveAt?: Date;
};
const Session: React.FC<React.PropsWithChildren<SessionProps>> = ({
  id,
  name,
  createdAt,
  finishedAt,
  clientName,
  clientLogoUri,
  lastActiveIp,
  lastActiveAt,
  isCurrent,
  children,
  deviceType,
}) => {
  const { t } = useTranslation();

  return (
    <Block className={styles.session}>
      <DeviceTypeIcon deviceType={deviceType || DeviceType.Unknown} />
      <div className={styles.container}>
        {isCurrent && (
          <Badge kind="success">{t("frontend.session.current_badge")}</Badge>
        )}
        <H6 className={styles.sessionName} title={id}>
          {name || id}
        </H6>
        <SessionMetadata weight="semibold">
          <Trans i18nKey="frontend.session.signed_in_date">
            Signed in <DateTime datetime={createdAt} />
          </Trans>
        </SessionMetadata>
        {!!finishedAt && (
          <SessionMetadata weight="semibold" data-finished={true}>
            <Trans i18nKey="frontend.session.finished_date">
              Finished <DateTime datetime={finishedAt} />
            </Trans>
          </SessionMetadata>
        )}
        {!!lastActiveAt && (
          <SessionMetadata>
            <LastActive lastActive={lastActiveAt} />
          </SessionMetadata>
        )}
        {!!lastActiveIp && <SessionMetadata>{lastActiveIp}</SessionMetadata>}
        {!!clientName && (
          <SessionMetadata>
            <ClientAvatar
              size="var(--cpd-space-4x)"
              name={clientName}
              logoUri={clientLogoUri}
            />{" "}
            <SessionMetadata weight="semibold" as="span">
              {clientName}
            </SessionMetadata>
          </SessionMetadata>
        )}
        {!!children && <div className={styles.sessionActions}>{children}</div>}
      </div>
    </Block>
  );
};

export default Session;
