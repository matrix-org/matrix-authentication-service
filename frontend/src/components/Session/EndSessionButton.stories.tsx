// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { Meta, StoryObj } from "@storybook/react";

import EndSessionButton from "./EndSessionButton";

const endSessionWithMockAsync = async (): Promise<void> => {
  console.log("End session submitted");
  await new Promise((resolve) => setTimeout(resolve, 1000));
  console.log("End session complete");
};

const meta = {
  title: "UI/Session/End session button",
  component: EndSessionButton,
  tags: ["autodocs"],
  args: {
    endSession: endSessionWithMockAsync,
    sessionCount: undefined,
  },
  argTypes: {
    endSession: { action: "end session confirmed!" },
    sessionCount: {
      control: "number",
    },
  },
} satisfies Meta<typeof EndSessionButton>;

export default meta;
type Story = StoryObj<typeof EndSessionButton>;

export const Basic: Story = {};

export const MultipleSessions: Story = {
  args: {
    sessionCount: 3,
  },
};
