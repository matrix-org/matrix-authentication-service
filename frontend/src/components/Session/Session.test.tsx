// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { create } from "react-test-renderer";
import { describe, expect, it, beforeAll } from "vitest";

import { mockLocale } from "../../test-utils/mockLocale";

import Session from "./Session";

describe("<Session />", () => {
  const defaultProps = {
    id: "session-id",
    createdAt: "2023-06-29T03:35:17.451292+00:00",
  };

  const finishedAt = "2023-06-29T03:35:19.451292+00:00";

  beforeAll(() => mockLocale());

  it("renders an active session", () => {
    const component = create(<Session {...defaultProps} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders a finished session", () => {
    const component = create(
      <Session {...defaultProps} finishedAt={finishedAt} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("uses session name when truthy", () => {
    const name = "test session name";
    const component = create(
      <Session {...defaultProps} finishedAt={finishedAt} name={name} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("uses client name when truthy", () => {
    const clientName = "Element";
    const component = create(
      <Session
        {...defaultProps}
        finishedAt={finishedAt}
        clientName={clientName}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
