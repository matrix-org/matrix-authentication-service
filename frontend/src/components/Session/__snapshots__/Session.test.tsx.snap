// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Session /> > renders a finished session 1`] = `
<div
  className="_block_17898c _session_634806"
>
  <svg
    aria-label="Unknown device type"
    className="_icon_e677aa"
    fill="currentColor"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M5 21c-.55 0-1.02-.196-1.413-.587A1.926 1.926 0 0 1 3 19V5c0-.55.196-1.02.587-1.413A1.926 1.926 0 0 1 5 3h14c.55 0 1.02.196 1.413.587.39.393.587.863.587 1.413v14c0 .55-.196 1.02-.587 1.413A1.926 1.926 0 0 1 19 21H5Zm0-2h14V5H5v14Z"
    />
    <path
      d="M11 10a1 1 0 1 1 1.479.878c-.31.17-.659.413-.94.741-.286.334-.539.8-.539 1.381a1 1 0 0 0 2 .006.3.3 0 0 1 .057-.085 1.39 1.39 0 0 1 .382-.288A3 3 0 1 0 9 10a1 1 0 1 0 2 0Zm1.999 3.011v-.004.005ZM12 17a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
    />
  </svg>
  <div
    className="_container_634806"
  >
    <h6
      className="_typography_yh5dq_162 _font-body-md-semibold_yh5dq_64 _sessionName_634806"
      title="session-id"
    >
      <a
        href="/sessions/session-id"
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        style={{}}
      >
        session-id
      </a>
    </h6>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
    >
      Signed in 
      <time
        dateTime="2023-06-29T03:35:17Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
      data-finished={true}
    >
      Finished 
      <time
        dateTime="2023-06-29T03:35:19Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
  </div>
</div>
`;

exports[`<Session /> > renders an active session 1`] = `null`;

exports[`<Session /> > renders ip address 1`] = `
<div
  className="_block_17898c _session_634806"
>
  <svg
    aria-label="Unknown device type"
    className="_icon_e677aa"
    fill="currentColor"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M5 21c-.55 0-1.02-.196-1.413-.587A1.926 1.926 0 0 1 3 19V5c0-.55.196-1.02.587-1.413A1.926 1.926 0 0 1 5 3h14c.55 0 1.02.196 1.413.587.39.393.587.863.587 1.413v14c0 .55-.196 1.02-.587 1.413A1.926 1.926 0 0 1 19 21H5Zm0-2h14V5H5v14Z"
    />
    <path
      d="M11 10a1 1 0 1 1 1.479.878c-.31.17-.659.413-.94.741-.286.334-.539.8-.539 1.381a1 1 0 0 0 2 .006.3.3 0 0 1 .057-.085 1.39 1.39 0 0 1 .382-.288A3 3 0 1 0 9 10a1 1 0 1 0 2 0Zm1.999 3.011v-.004.005ZM12 17a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
    />
  </svg>
  <div
    className="_container_634806"
  >
    <h6
      className="_typography_yh5dq_162 _font-body-md-semibold_yh5dq_64 _sessionName_634806"
      title="session-id"
    >
      <a
        href="/sessions/session-id"
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        style={{}}
      >
        session-id
      </a>
    </h6>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
    >
      Signed in 
      <time
        dateTime="2023-06-29T03:35:17Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
      data-finished={true}
    >
      Finished 
      <time
        dateTime="2023-06-29T03:35:19Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
    <p
      className="_typography_yh5dq_162 _font-body-sm-regular_yh5dq_40 _sessionMetadata_634806"
    >
      127.0.0.1
    </p>
    <p
      className="_typography_yh5dq_162 _font-body-sm-regular_yh5dq_40 _sessionMetadata_634806"
    >
       
      <span
        className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
      >
        Element
      </span>
    </p>
  </div>
</div>
`;

exports[`<Session /> > uses client name when truthy 1`] = `
<div
  className="_block_17898c _session_634806"
>
  <svg
    aria-label="Unknown device type"
    className="_icon_e677aa"
    fill="currentColor"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M5 21c-.55 0-1.02-.196-1.413-.587A1.926 1.926 0 0 1 3 19V5c0-.55.196-1.02.587-1.413A1.926 1.926 0 0 1 5 3h14c.55 0 1.02.196 1.413.587.39.393.587.863.587 1.413v14c0 .55-.196 1.02-.587 1.413A1.926 1.926 0 0 1 19 21H5Zm0-2h14V5H5v14Z"
    />
    <path
      d="M11 10a1 1 0 1 1 1.479.878c-.31.17-.659.413-.94.741-.286.334-.539.8-.539 1.381a1 1 0 0 0 2 .006.3.3 0 0 1 .057-.085 1.39 1.39 0 0 1 .382-.288A3 3 0 1 0 9 10a1 1 0 1 0 2 0Zm1.999 3.011v-.004.005ZM12 17a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
    />
  </svg>
  <div
    className="_container_634806"
  >
    <h6
      className="_typography_yh5dq_162 _font-body-md-semibold_yh5dq_64 _sessionName_634806"
      title="session-id"
    >
      <a
        href="/sessions/session-id"
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        style={{}}
      >
        session-id
      </a>
    </h6>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
    >
      Signed in 
      <time
        dateTime="2023-06-29T03:35:17Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
      data-finished={true}
    >
      Finished 
      <time
        dateTime="2023-06-29T03:35:19Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
    <p
      className="_typography_yh5dq_162 _font-body-sm-regular_yh5dq_40 _sessionMetadata_634806"
    >
      <img
        alt="Element"
        className="_avatar_39aee3"
        referrerPolicy="no-referrer"
        src="https://client.org/logo.png"
        style={
          {
            "--mas-avatar-size": "var(--cpd-space-4x)",
          }
        }
      />
       
      <span
        className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
      >
        Element
      </span>
    </p>
  </div>
</div>
`;

exports[`<Session /> > uses session name when truthy 1`] = `
<div
  className="_block_17898c _session_634806"
>
  <svg
    aria-label="Unknown device type"
    className="_icon_e677aa"
    fill="currentColor"
    height="1em"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M5 21c-.55 0-1.02-.196-1.413-.587A1.926 1.926 0 0 1 3 19V5c0-.55.196-1.02.587-1.413A1.926 1.926 0 0 1 5 3h14c.55 0 1.02.196 1.413.587.39.393.587.863.587 1.413v14c0 .55-.196 1.02-.587 1.413A1.926 1.926 0 0 1 19 21H5Zm0-2h14V5H5v14Z"
    />
    <path
      d="M11 10a1 1 0 1 1 1.479.878c-.31.17-.659.413-.94.741-.286.334-.539.8-.539 1.381a1 1 0 0 0 2 .006.3.3 0 0 1 .057-.085 1.39 1.39 0 0 1 .382-.288A3 3 0 1 0 9 10a1 1 0 1 0 2 0Zm1.999 3.011v-.004.005ZM12 17a1 1 0 1 0 0-2 1 1 0 0 0 0 2Z"
    />
  </svg>
  <div
    className="_container_634806"
  >
    <h6
      className="_typography_yh5dq_162 _font-body-md-semibold_yh5dq_64 _sessionName_634806"
      title="session-id"
    >
      <a
        href="/sessions/session-id"
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        style={{}}
      >
        test session name
      </a>
    </h6>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
    >
      Signed in 
      <time
        dateTime="2023-06-29T03:35:17Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
    <p
      className="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 _sessionMetadata_634806"
      data-finished={true}
    >
      Finished 
      <time
        dateTime="2023-06-29T03:35:19Z"
      >
        Thu, 29 Jun 2023, 03:35
      </time>
    </p>
  </div>
</div>
`;
